core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -321,14 +297,25 @@ private void sendHead() {
 
         private void sendTail() {
             assert state == State.REQUIRE_TAIL;
-            if (tail != null) {
-                final T tail = this.tail;
-                this.tail = null;
+            finalize(null);
+        }
+
+        private void finalize(@Nullable Throwable cause) {
+            final T tail;
+            try {
+                tail = finalizer.apply(cause);
+            } catch (Throwable ex) {
+                close0(ex);
```

Comment:
How about leaving a log and close with the original `cause`?

Reply:
Good idea.